### PR TITLE
Add at-sigatomic macro (closes #12309)

### LINKED ISSUE #12309: at-atomic macro as alternative to disable_sigint()
While working on VideoIO, I was unaware of `disable_sigint`, and created a macro which disables interrupts around sections of code:

```julia
macro atomic(code)
    quote
        Base.sigatomic_begin()
        ret = ($code)
        Base.sigatomic_end()
        ret
    end
end
```

Even after discovering `disable_sigint`, I like this a little better, especially for single ccalls (or in the case of VideoIO, wrapped ccalls):

```julia
@atomic av_freep(c)  # function which wraps ccall((:av_freep,libavutil),Void,(Ptr{Void},),ptr)
```

Is there interest in having this in Base?  If so, I'll submit a PR.

Cc: @stevengj

### PR DESCRIPTION
**Pull request by kmsquire**
cc: @stevengj @JeffBezanson @vtjnash

### CODE REVIEW

File: base/c.jl [hunk 1]
```diff
@@ -91,6 +91,10 @@ sigatomic_end() = ccall(:jl_sigatomic_end, Void, ())
 disable_sigint(f::Function) = try sigatomic_begin(); f(); finally sigatomic_end(); end
 reenable_sigint(f::Function) = try sigatomic_end(); f(); finally sigatomic_begin(); end
 
+macro sigatomic(code)
+    :(try sigatomic_begin(); $code; finally sigatomic_end(); end)
```
Comment (JeffBezanson) [inline base/c.jl]: Should be `$(esc(code))`.

Comment (kmsquire) [inline base/c.jl]: Fixed, thanks.

**Comment by vtjnash** [comment]
i think this should encourage use of the lambda form, rather than adding additional syntax:
```
help?> disable_sigint
search: disable_sigint

Base.disable_sigint(f::Function)

   Disable Ctrl-C handler during execution of a function, for calling
   external code that is not interrupt safe. Intended to be called
   using "do" block syntax as follows:

      disable_sigint() do
          # interrupt-unsafe code
          ...
      end
```

**Comment by kmsquire** [comment]
> i think this should encourage use of the lambda form, rather than adding additional syntax

Any suggestions on how it should do that?  

I find this slightly cleaner for one line `ccall`s, and I'm generally not a fan of the lambda forms because of an preconceived notion that they might be inefficient (because, well, they use lambdas).

Any reason why the lambda forms aren't used in GTK.jl?

(See also the discussion in #12309)

**Comment by vtjnash** [comment]
Gtk.jl has complicated needs due to COPY_STACKS interfering with event loop integration (prior to that it was using the lambda form of disable_sigint).

the syntax does look pretty sensible for ccall, although there's another experiment in making it zero-cost that I want to try first before adding this

inlining can (in the future) handle the lambda transformation to make it equivalent to this, so i'm not too convinced by that argument

**Comment by kmsquire** [comment]
> inlining can (in the future) handle the lambda transformation to make it equivalent to this, so i'm not too convinced by that argument

Okay.  There's been talk about various kinds of function inlining for a few years now.  I'm sure we're a lot closer than we were in 2012, but you should be able to understand why I would make that statement. ;-)

Looking forward to your zero-cost solution.

**Comment by yuyichao** [comment]
Now that we have jb/functions, is this still needed?

**Comment by kmsquire** [comment]
I argued above that this is cleaner for ccalls. @vtjnash said he had a zero-cost experiment he wanted to try first, and neither of us followed up.

**Comment by vtjnash** [comment]
no, you can close this with your branch `origin/yyc/threads/safepoint-signal`

**Comment by yuyichao** [comment]
Yep, the zero-cost approach is https://github.com/JuliaLang/julia/pull/16174 and I was doing my pre-PR opening search for related issues/prs.

FWIW, if there's still good reason to prefer a macro with #16174 and jb/functions, the try-catch is not necessary anymore.

**Comment by kmsquire** [comment]
#16174 looks nice!  I look forward to trying that with VideoIO sometime (although that package isn't in a very good state right now).

You can definitely close this and #12309 with the merge of #16174.

**Pull request closed without merging**